### PR TITLE
Fix catboost regression error for shap interaction values

### LINKED ISSUE #3416: BUG: compute interaction values for catboost
### Issue Description

When computing shap interaction values for a catboost model I get an error:
`AssertionError: The background dataset you provided does not cover all the leaves in the model, so TreeExplainer cannot run with the feature_perturbation="tree_path_dependent" option! Try providing a larger background dataset, no background dataset, or using feature_perturbation="interventional".`

But it would work with the shortcut using the C++ version of Tree SHAP as it is done for xgboost or for computing the shap values for catboost:
`model.get_feature_importance(catboost.Pool(X), fstr_type="ShapInteractionValues")`

### Minimal Reproducible Example

```python
import pandas as pd
import catboost
import shap

titanic_df = pd.read_csv("titanic.csv")
y = titanic_df["Survived"]
X = titanic_df.drop(columns=["Survived"])
model = catboost.CatBoostClassifier().fit(X,y)
explainer = shap.TreeExplainer(model)
explainer.shap_interaction_values(X)
```

### Traceback

_No response_

### Expected Behavior

_No response_

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [x] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [ ] I'd be interested in making a PR to fix this bug

### Installed Versions

Clone of the repo to have fix #3325 after release 0.43.0

### LINKED ISSUE #3457: BUG: interaction values for catboost regressor cannot be calculated
### Issue Description

When calculating the interaction values for a catboost regressor, an error is thrown.

### Minimal Reproducible Example

```python
import shap
import catboost

X, y = shap.datasets.california(n_points=50)
cat = catboost.CatBoostRegressor(depth=1, iterations=10).fit(X, y)
ex_cat = shap.TreeExplainer(cat)
e_cat = ex_cat(X, interactions=True)
```

### Traceback

```shell
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tobias/programming/github/shap/shap/explainers/_tree.py", line 248, in __call__
    v = self.shap_interaction_values(X)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/programming/github/shap/shap/explainers/_tree.py", line 569, in shap_interaction_values
    X, y, X_missing, flat_output, tree_limit, _ = self._validate_inputs(X, y, tree_limit, False)
                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tobias/programming/github/shap/shap/explainers/_tree.py", line 299, in _validate_inputs
    tree_limit = self.model.values.shape[0]
                 ^^^^^^^^^^^^^^^^^
AttributeError: 'TreeEnsemble' object has no attribute 'values'
```

### Expected Behavior

Not sure if a TreeEnsemble model should be generated for catboost. Would expect to calculate catboost interaction values without problems.

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

shap master (commit: 7bd33ff5dd5b7571cdec2dc097c7dc9facbf113b)

### PR DESCRIPTION
**Pull request by CloseChoice**
## Overview

Closes #3457
Closes #3416  

Description of the changes proposed in this pull request:
- fix catboost errors for regression model
- add tests for catboost regression (including interactions) and lightgbm
- use C++ shortcut implemented by catboost to calculate the interaction values (thanks @MLecardonnel)

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: shap/explainers/_tree.py [hunk 1]
```diff
@@ -565,6 +565,19 @@ def shap_interaction_values(self, X, y=None, tree_limit=None):
...
+            if type(X) != catboost.Pool:
+                X = catboost.Pool(X, cat_features=self.model.cat_feature_indices)
+            phi = self.model.original_model.get_feature_importance(data=X, fstr_type='ShapInteractionValues')
+            # note we pull off the last column and keep it as our expected_value
+            if len(phi.shape) == 4:
+                self.expected_value = getattr(self, "expected_value", [phi[0, i, -1, -1] for i in range(phi.shape[1])])
+                return [phi[:, i, :-1, :-1] for i in range(phi.shape[1])]
+            else:
+                self.expected_value = getattr(self, "expected_value", phi[0, -1, -1])
+                return phi[:, :-1, :-1]
```
Comment (connortann) [inline shap/explainers/_tree.py:583]: Do you think this diff should be it's own private method? e.g. 

```python
elif (self.model.model_type == "catboost") and (self.feature_perturbation == "tree_path_dependent"):
    assert tree_limit == -1, "tree_limit is not yet supported for CatBoost models!"
    return self._shap_interaction_values_catboost(X)
```

I'm in two minds but thought I'd suggest the idea. As a rather general observation, the shap codebase and particularly the tree explainer has some rather long functions with several chunky if/else statements. So, within reason we could look to identify blocks of code that could be broken down a little.

**Review by connortann — APPROVED** [review]
I added one minor suggestion above (up to you whether to implement), but otherwise this LGTM

File: shap/explainers/_tree.py (hunk 1 above)
Comment (CloseChoice) [reply shap/explainers/_tree.py:583]: I absolutely agree that these if/else statements make the functions clunky. But I would suggest that we gather a couple more tests and then tackle a general rewrite of the TreeExplainer.shap_values and TreeExplainer.shap_interaction_values methods (and some methods called in there, e.g. generate output, etc.). I strongly believe that there is a way better formulation than the one we have right now. 

So as for now, I would just leave it as it is.

**Comment by mayer79** [comment]
Great stuff, thanks!

**Pull request merged**